### PR TITLE
CUDA: Replace use of deprecated NVVM IR features, questionable constructs

### PR DESCRIPTION
**Pull request by gmarkall**
This change the following:

- Data layouts. Since CUDA 11.6 the preferred datalayout includes 128 bit integers. Since these are not supported on prior toolkits, we need to keep to old datalayout for CUDA 11.5 and earlier. Reference: [Data layouts NVVM IR spec for CUDA 11.6](https://docs.nvidia.com/cuda/archive/11.6.0/nvvm-ir-spec/index.html#data-layout).
- Address space conversions. The standard `addrspacecast` instruction is supported, and the intrinsics are deprecated. References:
  - [`addrspacecast` instruction](https://releases.llvm.org/7.0.1/docs/LangRef.html#addrspacecast-to-instruction)
  - [Supported conversion operations in CUDA 10.2](https://docs.nvidia.com/cuda/archive/10.2/nvvm-ir-spec/index.html#conversion-operations) - includes `addrspacecast` (referenced as this is the oldest supported version)
  - [Address space conversion intrinsics in CUDA 10.2](https://docs.nvidia.com/cuda/archive/10.2/nvvm-ir-spec/index.html#nvvm-intrin-addrsp) - they were even deprecated in 10.2!
- We ensure that kernels don't have the `noinline` attribute, as this causes NVVM to generate a warning
- We now declare that the IR version we supply is the same as the IR version of the current NVVM (rather than trying to guess based on whether we are using NVVM70 or not).
- Since this now causes NVVM to produce no known warnings, I also added raising of NVVM warnings if they are present (and added a test)

Commit messages contain some additional commentary on some small refactorings that were done as part of this work.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
There's a couple of issues I need to address:

- Now we need to know the NVVM version for the datalayout, the current implementation is always attempting to load NVVM.
- Something IR-related for NVVM34 (seen in gpuCI for CUDA 10.2 and 11.1)

**Comment by gmarkall** [comment]
gpu ci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Review by guilhermeleobas — DISMISSED** [review]
Hi @gmarkall, PR looks good to me. I only have one minor suggestion but don't feel obligated to address it.

File: numba/cuda/target.py [hunk 1]
```diff
@@ -295,9 +297,8 @@ def make_constant_array(self, builder, aryty, arr):
         gv.align = 2 ** (align - 1).bit_length()
 
         # Convert to generic address-space
-        conv = nvvmutils.insert_addrspace_conv(lmod, ir.IntType(8), addrspace)
-        addrspaceptr = gv.bitcast(ir.PointerType(ir.IntType(8), addrspace))
-        genptr = builder.call(conv, [addrspaceptr])
+        ptrty = ir.PointerType(ir.IntType(8))
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/target.py:300]: Do you mind if I don't? It's used twice - in one place it's logically used as a pointer to a char, and in another as a generic "void" pointer, with names reflecting their usage - so making them both `i8p` strips out a little semantic information for the reader, even if the types would be the same.

**Comment by gmarkall** [comment]
@guilhermeleobas Many thanks for the review - I've now merged master and responded to the comment above. Note I had to make a couple of edits to resolve the conflict, which are in keeping with the original changes in the PR: https://github.com/numba/numba/pull/8290/commits/f52991bb413cf7501fdce28fc13827bc960609ab

@esc Could this have a CUDA buildfarm run please?

File: numba/cuda/target.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cuda/target.py:300]: No, that's completely fine.

[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/target.py:300]: Many thanks!

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, couple of minor comments to resolve else looks good.

File: numba/cuda/cudadrv/nvvm.py [hunk 2]
```diff
@@ -893,20 +905,14 @@ def set_cuda_kernel(lfunc):
     nmd = cgutils.get_or_insert_named_metadata(mod, 'nvvm.annotations')
     nmd.add(md)
 
+    # Marking a kernel 'noinline' causes NVVM to generate a warning, so remove
+    # it if it is present.
+    lfunc.attributes.discard('noinline')
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/cudadrv/nvvm.py:910]: I don't think so. I don't think it actually matters if `noinline` is present, but a quirk of NVVM causes it to complain when `noinline` is present on a kernel.

File: numba/cuda/cudadrv/nvvm.py [hunk 3]
```diff
@@ -893,20 +905,14 @@ def set_cuda_kernel(lfunc):
...
-    i32 = ir.IntType(32)
-    if NVVM().is_nvvm70:
-        # NVVM IR 1.6, DWARF 3.0
-        ir_versions = [i32(1), i32(6), i32(3), i32(0)]
-    else:
-        # NVVM IR 1.1, DWARF 2.0
-        ir_versions = [i32(1), i32(2), i32(2), i32(0)]
-
+    # We specify the IR version to match the current NVVM's IR version
+    ir_versions = [ir.IntType(32)(v) for v in NVVM().get_ir_version()]
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/cudadrv/nvvm.py:916]: No, but as it happens exactly 4 times and fit within one line I just squeezed it in there anyway. I don't mind how this looks either way.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
@stuartarchibald Many thanks for the review - I have responded to both comments.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_158`.

Passed (on all non-CUDA-10.2 builds as expected since #8287)!

**Pull request merged**